README.md
Quran Finder
============

Limnoria/Supybot plugin that brings verses from Quran using a website API

It was developed by a member of ##islam channel in freenode.net IRC network.

How to Use:

```
!quran <surah> <ayah> <lang>
```
returns ayah number \<ayah\> of surah number \<surah\> in \<lang\> language.

```
!quran 1 1 en
```
Returns 
```
1,1: Praise be to Allah, Lord of the Worlds,
```

The bot can obtain translation of the Quran for different languages. (ar, en, fa, tr) are the main ones. You can get extra translations and other data sources by using the data source key provided by Global Quran API. For more information see the [Wiki](https://github.com/SafaAlfulaij/QuranFinder/wiki)

One example is:
```
!quran 1 1 en.hilali
```
which will provide the Hilali English translation of the Quran.

_NOTE: the '!' character is just prefix command character of the bot and may be different depending how the bot was configured_

## License
This program is free software under GNU General Public License 3 or any later version and COMES WITHOUT AND WARRANTY, for more information see LICENSE.

## To-Do
- [ ] Don't send a message for each ayah and split messages, instead use supybot's more
- [ ] Use config.py for the maximum difference between two verses(by default) it is 5


CHANGELOG.md
# Change Log

Version 0.4.2 - (12 May 2016)
==========
 - Workaround to avoid non-latin text getting omitted [Limnoria bug 1232] (https://github.com/ProgVal/Limnoria/issues/1232)
 - Info about the implemented support for API keys were added to Readme file.
 - Removed version line from plugin.py file to avoid double changing when accepting a pull request.


Version 0.4.1 - (1 May 2016)
==========
 - Added links to Wiki in Bot messages for help (e.g: API_codes).
 - Enabled Token again.

Version 0.4 - (30 Apri 2016)
==========
 - English is the default. No need to write `en`.
 - Support for <lang>.<translation> so now you can most of the API data. (e.g: ar.jalalayn, en.hilali) from http://api.globalquran.com/quran

Version 0.3 - (21 Apri 2016)
==========
 - Added Turkish language translation
 - Changed English translation to Saheeh International because of it's simplicity

Version 0.2 - (20 Apri 2016)
==========
 - Change the API to globalquran.com, privious one had Arabic and English only, it's also limited to one old English translation (Pickthall)
 - Used Muhammad Taqi-ud-Din al-Hilali and Muhammad Muhsin Khan translation


Version 0.1 (27 February 2016)
==========
 - First bot release


test.py

# Copyright 2016 Safa AlFulaij <safa1996alfulaij@gmail.com>
#
# This file is part of QuranFinder.
#
# QuranFinder is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# QuranFinder is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with QuranFinder.  If not, see <http://www.gnu.org/licenses/>.


from supybot.test import *


class QuranFinderTestCase(PluginTestCase):
    plugins = ('QuranFinder',)


# vim:set shiftwidth=4 tabstop=4 expandtab textwidth=79:


config.py

# Copyright 2016 Safa AlFulaij <safa1996alfulaij@gmail.com>
#
# This file is part of QuranFinder.
#
# QuranFinder is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# QuranFinder is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with QuranFinder.  If not, see <http://www.gnu.org/licenses/>.


import supybot.conf as conf
import supybot.registry as registry
try:
    from supybot.i18n import PluginInternationalization
    _ = PluginInternationalization('QuranFinder')
except:
    # Placeholder that allows to run the plugin on a bot
    # without the i18n module
    _ = lambda x: x


def configure(advanced):
    # This will be called by supybot to configure this module.  advanced is
    # a bool that specifies whether the user identified themself as an advanced
    # user or not.  You should effect your configuration by manipulating the
    # registry as appropriate.
    from supybot.questions import expect, anything, something, yn
    QuranFinder = conf.registerPlugin('QuranFinder', True)
    if yn("""Split long verses?""", default=True):
        QuranFinder.splitMessages.setValue(True)
    else:
        QuranFinder.splitMessages.setValue(False)

QuranFinder = conf.registerPlugin('QuranFinder')
# This is where your configuration variables (if any) should go.  For example:
conf.registerGlobalValue(QuranFinder, 'splitMessages',
    registry.Boolean(True, _("""Set to split long verses.""")))
conf.registerGlobalValue(QuranFinder, 'splitMessagesAt',
    registry.PositiveInteger(350, _("""Set where to split long verses.""")))


# vim:set shiftwidth=4 tabstop=4 expandtab textwidth=79:


__init__.py

# Copyright 2016 Safa AlFulaij <safa1996alfulaij@gmail.com>
#
# This file is part of QuranFinder.
#
# QuranFinder is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# QuranFinder is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with QuranFinder.  If not, see <http://www.gnu.org/licenses/>.


"""
QuranFinder: Brings verses from Quran
"""

import supybot
import supybot.world as world

# Use this for the version of this plugin.  You may wish to put a CVS keyword
# in here if you're keeping the plugin in CVS or some similar system.
__version__ = ""

# XXX Replace this with an appropriate author or supybot.Author instance.
__author__ = supybot.authors.unknown

# This is a dictionary mapping supybot.Author instances to lists of
# contributions.
__contributors__ = {}

# This is a url where the most recent plugin package can be downloaded.
__url__ = ''

from . import config
from . import plugin
from imp import reload
# In case we're being reloaded.
reload(config)
reload(plugin)
# Add more reloads here if you add third-party modules and want them to be
# reloaded when this plugin is reloaded.  Don't forget to import them as well!

if world.testing:
    from . import test

Class = plugin.Class
configure = config.configure


# vim:set shiftwidth=4 tabstop=4 expandtab textwidth=79:


plugin.py
# Copyright 2016 Safa AlFulaij <safa1996alfulaij@gmail.com>
#
# This file is part of QuranFinder.
#
# QuranFinder is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# QuranFinder is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with QuranFinder.  If not, see <http://www.gnu.org/licenses/>.

import supybot.utils as utils
from supybot.commands import *
import supybot.plugins as plugins
import supybot.ircutils as ircutils
import supybot.callbacks as callbacks
try:
    from supybot.i18n import PluginInternationalization
    _ = PluginInternationalization('QuranFinder')
except ImportError:
    # Placeholder that allows to run the plugin on a bot
    # without the i18n module
    _ = lambda x: x

API_URL = "http://api.globalquran.com/ayah/" #verse:ayah/quranID
# Set languages and translations the bot will accept
quranID = {"ar" : "quran-simple", "en" : "en.sahih", "tr" : "tr.yazir", "fa" : "fa.fooladvand"}

TOKEN = "the token" # Token obtained from http://docs.globalquran.com

import requests

class qdata():
    def __init__(self, chapter, ayah, lang):
        if(lang == None):
            lang = "en"
        if(len(lang) == 2):
            if(lang not in quranID):
                raise ValueError("Only " + " ,".join(quranID) + " languages are supported using two letters code. Maybe you would like to use a translation/tafsir code instead, Check: https://git.io/vwMz4 for a list of avalible sources.")
            lang = quranID[lang]

        if (chapter > 114 or chapter < 1):
            raise ValueError("Invalid Surah number.")

        json = self.requestData(chapter, ayah, lang)
        self.parseResponse(json)

        if (int(self.SurahNumber) != chapter): # If the ayah number is bigger
        # than the ayahs in the surah, the API jumps to another surah.
            raise ValueError("Invalid Ayah number.")

    def requestData(self, chapter, ayah, lang):
        url = API_URL + str(chapter) + ":" + str(ayah) + "/" + lang
        request = requests.get(url, params = {'key' : TOKEN})
        json = request.json()

        # the ID differs for each verse. So there is no static key to call in the main json.
        for quran in json:
            json = json[quran]
            for quranVer in json: # the QuranID in the json. Here we used quranVar to avoid conflict with the quranID above.
                json = json[quranVer]
                for ID in json:
                    json = json[ID]

        return json

    def parseResponse(self, json):
        self.SurahNumber = json["surah"]
        self.ayahNumber = json["ayah"]
        self.ayahText = json["verse"]
class utilities():
    def arg2list(ayat): # TODO: better name for this function
        MAX_DIFF = 5
        try:
            return [int(ayat)]
        except ValueError:
            t = ayat.split('-') 
            start, end = int(t[0]), int(t[1])
            if (end - start) not in range(1, MAX_DIFF + 1):
                raise ValueError("Out of range(maximum is 5) or start ayah is" +
                " bigger than end ayah.") #TODO: a better message please
            else:
                return list(range(start, end + 1))
            


class QuranFinder(callbacks.Plugin):
    """This plugin gets verse and ayah number and sends you the ayah using a web API."""

    def __init__(self, irc):
         self.__parent = super(QuranFinder, self)
         self.__parent.__init__(irc)

    def quran(self, irc, msg, args, surah, ayat, lang):
        """<surah> <ayah/ayat> <lang>

        returns ayah number <ayah> of surah number <surah> in <lang> language or translation or tafsir. for more information visit: https://git.io/vwMz9
        """
        
        try:
            list_of_ayat = utilities.arg2list(ayat)
        except ValueError as e:
            irc.error(str(e))
            return

        try:
            ayat_list = []
            for ayah in list_of_ayat:
                data = qdata(surah, ayah, lang)
                ayat_list.extend([str(data.SurahNumber) + ":" + str(data.ayahNumber)
                           + ", " + str(data.ayahText)])
        except ValueError as e:
            irc.error(str(e))
            return
        except (KeyError, TypeError) as e: #TypeError incase requesting a audio version. The json would contain a list so qdata would raise a TypeError
            irc.error("Wrong translation code or broken API.") 
            return
        for ayah in ayat_list:

            if self.registryValue('splitMessages'):
                ircMsgBytes = ayah.encode('utf-8')
                while len(ircMsgBytes) > 350:
                    splitPoint = ircMsgBytes[0:351].rfind(' '.encode('utf-8'))
                    irc.reply(ircMsgBytes[0:splitPoint].decode('utf-8').strip())
                    ircMsgBytes = ircMsgBytes[splitPoint:]
                irc.reply(ircMsgBytes.decode('utf-8').strip())
            else:
                irc.reply(ayah)

    quran = wrap(quran, ["int", "something", optional("something")])




Class = QuranFinder


# vim:set shiftwidth=4 softtabstop=4 expandtab textwidth=79:
